### PR TITLE
✨ Add volume size limits for user services 

### PR DESCRIPTION
**Pull request by GitHK**
## ⚠️ devops

- Added new env var `DYNAMIC_SIDECAR_ENABLE_VOLUME_LIMITS` to enable the limitation of service's disk space.
  - https://github.com/ITISFoundation/osparc-ops-environments/pull/154
  - https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/merge_requests/4

## What do these changes do?

Certain services can now limit the amount of data on each volume a user can produce. This is defined in the service's labels.
The `PathMappingsLabel` was extended to support this.

**NOTE:** to limit the root file system we need to publish a compose spec version 2.3 and for each container in the compose spec specify something like `"storage_opt": {"size": "10M"} }`

- [x] check if node where sidecar is running supports quotas before applying this. This depends on the file system and some configuration, for details check https://github.com/ITISFoundation/osparc-simcore/issues/3741
- [x] expand example .yaml service spec with new labels after https://github.com/ITISFoundation/osparc-simcore/pull/3606 is merged

## Related issue/s

- fixes https://github.com/ITISFoundation/osparc-simcore/issues/3741
- https://github.com/ITISFoundation/osparc-issues/issues/638

## Checklist

- [x] Unit tests for the changes exist

### CODE REVIEW

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/api/docker.py
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/api/docker.py:8]: Sorry, but this is a really strange entrypoint? 
- it does not identify any resource
- ``supported`` is not an operation (i.e. custom operation are suppose to "replace" the http operation POST)
- you use the 404 (not found) to respond that quotas are not supported! 

Some suggestions? You want to run a check to determine whether volumes with are supported or not
- user an existing API resource as path ``docker/volumes`` 
- create an operation e.g. ``check`` or ``analyze`` , i.e. ``POST docker/volumes:check`` 
- respond with a body that has resulting information e.g.
```json
{
  "are_quotas_supported" : false
}
```
- This in addition is reusable for other properties you want to check/analyze about volumes
- Use http error codes to indicate states like it could not be "checked" (e.g. if docker client fails I would use a 503 for it)

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py:257]: you should distinguish between an error because quotas really cannot be created and a completely different error thrown by docker client/daemon

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py:260]: what if delete raises?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_thin.py
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_thin.py:228]: I think that for custom operations (i.e. those that doe not correspond to CRUD) it is better to name the functions in the ``_thin`` interface with the operation itself.

For instance imaging a resource ``/foo``. The thin interface would be
```python
# CRUD operations
async def list_foo(...):
  # GET /foo

async def get_foo(...): 
    # GET /foo/*

async def create_foo(...): 
    # POST /foo/*

async def update_foo(): 
    # PUT /foo/*

async def delete_foo(...): 
    # DELETE /foo/*

# CUSTOM OPERATORS
async def check_foo(): 
    # PATCH /foo/*:check

```

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py:177]: why not put this outside?
```python
if has_quota_support:
   _update_service_quotas(service_spec)
```

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:136]: Q: Did you check if ByteSize is able to parse those?

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:163]: maybe add what docker sizes are. or a reference at least.

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:179]: :+1:

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:210]: what is "/s" for? I do not get why the ALLOWED_DOCKER_SIZE is in the state_paths? what is the idea?

File: packages/models-library/tests/test_service_settings_labels.py
Comment (sanderegg) [inline packages/models-library/tests/test_service_settings_labels.py:156]: minor: maybe using the faker for the path, or something like "/a/valid/path" is more helpful to understand the path is valid

**Review by sanderegg — COMMENTED** [review]
nice job!
Some minor comments.

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (GitHK) [reply packages/models-library/src/models_library/service_settings_labels.py]: outdated, using ByteSize

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (pcrespov) [inline packages/models-library/src/models_library/service_settings_labels.py:188]: The same validation can be simply achieved simply adding it in the type annotation as

```python
volume_size_limits: Optional[dict[str, ByteSize]] 
```

NOTE that ``ByteSize`` is actually a ``str`` with class methods for validation!

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py:262]: Messages to identify an error is dangerous since those are for humans and might change. What about status code or any other error code they might provide?

- Can you pls add the link to where is this message documented? This way we can check if the message changed?
- Can you add this error in a constant and use it in the tests as well ?  ``DOCKER_QUOTAS_ERROR_MSG= "create {volume_name}: quota size requested but no quota support"``

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py
Comment (pcrespov) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py:269]: THOUGHT: you want to check whether your volume has quotas and for that **you perform a test**. If the test fails in a very specific way, you conclude it has not quotas. OK.
Now if the test fails for other reason or during cleanup (i.e. docker_volume.delete()), it raises an exception that probably ends up as a 500 that reaches the client.  The client should interpret that as "cannot know whether it has quotas or not".

If it raises **during cleanup**  (i.e. ``docker_volume.delete()``) I find that the client got the the wrong interpretation since the actual test you were intended to do succeeded in the first place. What went wrong is something else that has nothing to do with the answer that the client is interested in.

IMO that exception should be handled in place. Moreoever, i would do it in a background task as a ``on_cleanup`` event independent from the request.

File: services/dynamic-sidecar/tests/unit/conftest.py
Comment (pcrespov) [inline services/dynamic-sidecar/tests/unit/conftest.py:212]: - Is it really a 404 what is thrown when you cannot create a volume??
- Can you pls use DOCKER_QUOTAS_ERROR_MSG described above in this test?

**Review by pcrespov — APPROVED** [review]
Some extra comments. Please have a extra look before merging.
thx

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (sanderegg) [reply packages/models-library/src/models_library/service_settings_labels.py:188]: @pcrespov ByteSize is not a string, but bytes or int actually.

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:171]: Q: why not having a dict[str, ByteSize]???

File: packages/models-library/src/models_library/service_settings_labels.py
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:230]: :+1:

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py:491]: :+1:

**Review by sanderegg — APPROVED** [review]
Cool! Some last comments. good job.

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py]: Let's see how this will behave, moved it to a background task.

**Comment by GitHK** [comment]
What is blocking this PR?
Quota support needs to be determined on a "per node basis". We wanted to support a system where it enabled and disables itself based on the current status of the node.
Since the volume is created at the same time as the dy-sidecar and it is not possible to figure out in advance where swarm places the sidecar (nor do we want to change this), it is not possible to use this feature where it figure out if quota is supported or not.

As an alternative we need to run xfs on all nodes and have a setting enabling or disabling this (since some development machine might not have an xfs file system).

**Comment by GitHK** [comment]
Unblocked by using an env var to enable limits per deployment.

**Pull request merged**
